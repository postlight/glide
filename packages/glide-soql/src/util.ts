export const IS_EMPTY = Symbol("glide.isEmpty");

export interface Fragment {
  [IS_EMPTY](): boolean;
  toString(): string;
}

export function notEmpty(input: Fragment): boolean {
  return !isEmpty(input);
}

export function isEmpty(input: Fragment): boolean {
  return input[IS_EMPTY]();
}
