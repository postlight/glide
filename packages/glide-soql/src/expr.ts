import * as ops from "./ops";
import { IS_EMPTY, Fragment } from "./util";

const SINGLE_QUOTE = /'/gm;

export type Expr = Bool | Cmp;
export type Field = string | Fn;
export type Scalar = bigint | boolean | null | number | string;

export class Bool implements Fragment {
  private readonly value: Expr[];

  private constructor(private readonly operator: ops.Bool, value: Expr[]) {
    this.value = value.reduce<Expr[]>((acc, item) => {
      if (item instanceof Cmp || operator !== item.operator) {
        acc.push(item);
        return acc;
      }

      return acc.concat(item.value);
    }, []);
  }

  static and(...value: Expr[]): Bool {
    return new Bool(ops.Bool.And, value);
  }

  static or(...value: Expr[]): Bool {
    return new Bool(ops.Bool.Or, value);
  }

  [IS_EMPTY](): boolean {
    return this.value.length === 0;
  }

  toString(): string {
    return this.value
      .map(expr => (expr instanceof Bool ? `(${expr})` : `${expr}`))
      .join(` ${this.operator} `);
  }
}

export class Cmp implements Fragment {
  private constructor(
    private readonly field: Field,
    private readonly operator: ops.Cmp,
    private readonly expression: Scalar | Scalar[],
  ) {}

  static eq(field: Field, expression: Scalar): Cmp {
    return new Cmp(field, ops.Cmp.Eq, expression);
  }

  static gt(field: Field, expression: Scalar): Cmp {
    return new Cmp(field, ops.Cmp.Gt, expression);
  }

  static gte(field: Field, expression: Scalar): Cmp {
    return new Cmp(field, ops.Cmp.Gte, expression);
  }

  static in(field: Field, expression: Scalar[]): Cmp {
    return new Cmp(field, ops.Cmp.In, expression);
  }

  static like(field: Field, expression: string): Cmp {
    return new Cmp(field, ops.Cmp.Like, expression);
  }

  static lt(field: Field, expression: Scalar): Cmp {
    return new Cmp(field, ops.Cmp.Lt, expression);
  }

  static lte(field: Field, expression: Scalar): Cmp {
    return new Cmp(field, ops.Cmp.Lte, expression);
  }

  static ne(field: Field, expression: Scalar): Cmp {
    return new Cmp(field, ops.Cmp.Ne, expression);
  }

  static notIn(field: Field, expression: Scalar[]): Cmp {
    return new Cmp(field, ops.Cmp.NotIn, expression);
  }

  [IS_EMPTY](): boolean {
    return false;
  }

  toString(): string {
    let expression: string;

    if (Array.isArray(this.expression)) {
      expression = `(${this.expression.map(stringifyScalar).join(", ")})`;
    } else {
      expression = stringifyScalar(this.expression);
    }

    return `${this.field} ${this.operator} ${expression}`;
  }
}

export class Fn implements Fragment {
  private constructor(private readonly ident: string, private readonly field?: string) {}

  static avg(field: string): Fn {
    return new Fn("AVG", field);
  }

  static calendarMonth(field: string): Fn {
    return new Fn("CALENDAR_MONTH", field);
  }

  static calendarQuarter(field: string): Fn {
    return new Fn("CALENDAR_QUARTER", field);
  }

  static calendarYear(field: string): Fn {
    return new Fn("CALENDAR_YEAR", field);
  }

  static convertTimezone(field: string): Fn {
    return new Fn("CONVERT_TIMEZONE", field);
  }

  static count(field?: string): Fn {
    return new Fn("COUNT", field);
  }

  static countDistinct(field?: string): Fn {
    return new Fn("COUNT_DISTINCT", field);
  }

  static dayInMonth(field: string): Fn {
    return new Fn("DAY_IN_MONTH", field);
  }

  static dayInWeek(field: string): Fn {
    return new Fn("DAY_IN_WEEK", field);
  }

  static dayInYear(field: string): Fn {
    return new Fn("DAY_IN_YEAR", field);
  }

  static dayOnly(field: string): Fn {
    return new Fn("DAY_ONLY", field);
  }

  static fiscalMonth(field: string): Fn {
    return new Fn("FISCAL_MONTH", field);
  }

  static fiscalQuarter(field: string): Fn {
    return new Fn("FISCAL_QUARTER", field);
  }

  static fiscalYear(field: string): Fn {
    return new Fn("FISCAL_YEAR", field);
  }

  static hourInDay(field: string): Fn {
    return new Fn("HOUR_IN_DAY", field);
  }

  static max(field: string): Fn {
    return new Fn("MAX", field);
  }

  static min(field: string): Fn {
    return new Fn("MIN", field);
  }

  static sum(field: string): Fn {
    return new Fn("SUM", field);
  }

  static weekInMonth(field: string): Fn {
    return new Fn("WEEK_IN_MONTH", field);
  }

  static weekInYear(field: string): Fn {
    return new Fn("WEEK_IN_YEAR", field);
  }

  [IS_EMPTY](): boolean {
    return false;
  }

  toString(): string {
    return `${this.ident}(${this.field ? this.field : ""})`;
  }
}

function stringifyScalar(input: Scalar): string {
  switch (typeof input) {
    case "boolean":
      return `${input}`.toUpperCase();
    case "string":
      return `'${input.replace(SINGLE_QUOTE, "\\'")}'`;
    default:
      return input == null ? "null" : String(input);
  }
}
