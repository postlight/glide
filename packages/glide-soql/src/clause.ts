import { Bool, Cmp, Field } from "./expr";
import { IS_EMPTY, Fragment, isEmpty, notEmpty } from "./util";

export type Direction = "ASC" | "DESC";

export class GroupBy implements Fragment {
  constructor(private readonly fields: Field[] = [], private readonly having?: Bool | Cmp) {}

  [IS_EMPTY](): boolean {
    return this.fields.length === 0;
  }

  toString(): string {
    let clause = `GROUP BY ${this.fields.join(", ")}`;

    if (this.having && notEmpty(this.having)) {
      clause += ` HAVING ${this.having}`;
    }

    return clause;
  }
}

export class Limit implements Fragment {
  private readonly size?: number | void;

  constructor(size?: bigint | number) {
    this.size = tryIntoSafeInteger(size);
  }

  [IS_EMPTY](): boolean {
    return this.size == null;
  }

  toString(): string {
    return `LIMIT ${this.size}`;
  }
}

export class Offset implements Fragment {
  private readonly size?: number | void;

  constructor(size?: bigint | number) {
    this.size = tryIntoSafeInteger(size);
  }

  [IS_EMPTY](): boolean {
    return this.size == null;
  }

  toString(): string {
    return `OFFSET ${this.size}`;
  }
}

export class OrderBy implements Fragment {
  private readonly direction: Direction = "ASC";

  constructor(private readonly fields: Field[] = [], direction?: Direction) {
    this.direction = intoDirection(direction);
  }

  [IS_EMPTY](): boolean {
    return this.fields.length === 0;
  }

  toString(): string {
    return `ORDER BY ${this.fields.join(", ")} ${this.direction}`;
  }
}

export class Where implements Fragment {
  constructor(private readonly condition?: Bool | Cmp) {}

  [IS_EMPTY](): boolean {
    return this.condition == null || isEmpty(this.condition);
  }

  toString(): string {
    return `WHERE ${this.condition}`;
  }
}

function intoDirection(value: unknown): Direction {
  return value === "DESC" ? "DESC" : "ASC";
}

function tryIntoSafeInteger(value: unknown): number | void {
  const num = Number(value);
  return Number.isSafeInteger(num) && num >= 0 ? num : undefined;
}
