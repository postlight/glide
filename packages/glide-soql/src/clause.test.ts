import { GroupBy, Limit, Offset, OrderBy, Where } from "./clause";
import { Bool, Cmp, Fn } from "./expr";
import { isEmpty } from "./util";

describe("GroupBy", () => {
  test("#@@isEmpty()", () => {
    expect(isEmpty(new GroupBy())).toBe(true);
    expect(isEmpty(new GroupBy(["FirstName"]))).toBe(false);
  });

  test("#toString()", () => {
    const condition = Bool.and(Cmp.gt(Fn.count("Name"), 100), Cmp.eq("LeadSource", "Phone"));

    expect(`${new GroupBy(["LeadSource"])}`).toMatchSnapshot();
    expect(`${new GroupBy(["LeadSource"], condition)}`).toMatchSnapshot();
  });
});

describe("Limit", () => {
  test("#@@isEmpty()", () => {
    expect(isEmpty(new Limit())).toBe(true);
    expect(isEmpty(new Limit(10))).toBe(false);
  });

  test("#toString()", () => {
    expect(`${new Limit(10)}`).toMatchSnapshot();
  });
});

describe("Offset", () => {
  test("#@@isEmpty()", () => {
    expect(isEmpty(new Offset())).toBe(true);
    expect(isEmpty(new Offset(10))).toBe(false);
  });

  test("#toString()", () => {
    expect(`${new Offset(10)}`).toMatchSnapshot();
  });
});

describe("OrderBy", () => {
  test("#@@isEmpty()", () => {
    expect(isEmpty(new OrderBy())).toBe(true);
    expect(isEmpty(new OrderBy(["FirstName"]))).toBe(false);
  });

  test("#toString()", () => {
    expect(`${new OrderBy(["FirstName"])}`).toMatchSnapshot();
    expect(`${new OrderBy(["FirstName"], "ASC")}`).toMatchSnapshot();
    expect(`${new OrderBy(["FirstName"], "DESC")}`).toMatchSnapshot();
  });
});

describe("Where", () => {
  const condition = Cmp.eq("EmailAddress", "hello@postlight.com");

  test("#@@isEmpty()", () => {
    expect(isEmpty(new Where())).toBe(true);
    expect(isEmpty(new Where(condition))).toBe(false);
  });

  test("#toString()", () => {
    expect(`${new Where(condition)}`).toMatchSnapshot();
  });
});
