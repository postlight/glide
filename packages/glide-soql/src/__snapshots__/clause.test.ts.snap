// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GroupBy #toString() 1`] = `"GROUP BY LeadSource"`;

exports[`GroupBy #toString() 2`] = `"GROUP BY LeadSource HAVING COUNT(Name) > 100 AND LeadSource = 'Phone'"`;

exports[`Limit #toString() 1`] = `"LIMIT 10"`;

exports[`Offset #toString() 1`] = `"OFFSET 10"`;

exports[`OrderBy #toString() 1`] = `"ORDER BY FirstName ASC"`;

exports[`OrderBy #toString() 2`] = `"ORDER BY FirstName ASC"`;

exports[`OrderBy #toString() 3`] = `"ORDER BY FirstName DESC"`;

exports[`Where #toString() 1`] = `"WHERE EmailAddress = 'hello@postlight.com'"`;
